iceoryx_posh/source/popo/active_call_set.cpp

```diff
@@ -0,0 +1,223 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/popo/active_call_set.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/event_listener.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/event_notifier.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/cxx/helplets.hpp"
+
+namespace iox
+{
+namespace popo
+{
+ActiveCallSet::ActiveCallSet() noexcept
+    : ActiveCallSet(runtime::PoshRuntime::getInstance().getMiddlewareEventVariable())
+{
+}
+
+ActiveCallSet::ActiveCallSet(EventVariableData* eventVariable) noexcept
+    : m_eventVariable(eventVariable)
+    , m_eventListener(*eventVariable)
+{
+    m_thread = std::thread(&ActiveCallSet::threadLoop, this);
+}
+
+ActiveCallSet::~ActiveCallSet()
+{
+    m_wasDtorCalled.store(true, std::memory_order_relaxed);
+    m_eventListener.destroy();
+
+    m_thread.join();
+    m_eventVariable->m_toBeDestroyed.store(true, std::memory_order_relaxed);
+}
+
+cxx::expected<uint64_t, ActiveCallSetError>
+ActiveCallSet::addEvent(void* const origin,
+                        const uint64_t eventType,
+                        const uint64_t eventTypeHash,
+                        CallbackRef_t<void> callback,
+                        TranslationCallbackRef_t translationCallback,
+                        const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept
+{
+    std::lock_guard<std::mutex> lock(m_addEventMutex);
+
+    for (uint64_t i = 0U; i < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++i)
+    {
+        if (m_events[i]->isEqualTo(origin, eventType, eventTypeHash))
+        {
+            return cxx::error<ActiveCallSetError>(ActiveCallSetError::EVENT_ALREADY_ATTACHED);
+        }
+    }
+
+    uint32_t index = 0U;
+    if (!m_indexManager.pop(index))
+    {
+        return cxx::error<ActiveCallSetError>(ActiveCallSetError::ACTIVE_CALL_SET_FULL);
+    }
+
+    m_events[index]->init(index, origin, eventType, eventTypeHash, callback, translationCallback, invalidationCallback);
+    return cxx::success<uint64_t>(index);
+}
+
+void ActiveCallSet::removeEvent(void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept
+{
+    for (uint32_t index = 0U; index < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++index)
+    {
+        if (m_events[index]->resetIfEqualTo(origin, eventType, eventTypeHash))
+        {
+            m_indexManager.push(index);
+            break;
+        }
+    }
+}
+
+uint64_t ActiveCallSet::size() const noexcept
+{
+    return m_indexManager.size();
```

Comment:
~~I'm a little bit confused. I assumed `size()` returns the number of events the `ActiveCallSet` currently holds, but since `removeEvent` pushes to `m_indexManager`, this is the size of the available events to add, right? Wouldn't it make sense to call this `remainingCapacity` to eliminate the ambiguity?~~

Okay, after having look at `IndexManager` this is correct, but `size` behaves counterintuitive to `push` and `pop`. Please consider renaming them, e.g. `acquire` instead of `pop` and `release` instead of `push`. Or maybe `IndexManager::size` could be renamed to something like `indicesInUse`